### DISCUSSION TITLE
Integrating quantum computing resistant password hashing algorithm to Keycloak

### DISCUSSION THREAD

**Discussion by hmirheydari** [keycloak/keycloak#32007] [Ideas]
The advent of quantum computing poses a significant threat to the security of traditional cryptographic systems, including password hashing algorithms. While still in its nascent stages, quantum computing has the potential to drastically accelerate brute-force attacks, rendering current hashing methods obsolete.

This discussion aims to explore the feasibility and benefits of integrating a quantum-resistant password hashing algorithm into Keycloak. We will discuss the following key points:

Understanding the Quantum Threat: A brief overview of quantum computing and its implications for password security.
Post-Quantum Cryptography: Exploring existing and emerging post-quantum cryptographic algorithms suitable for password hashing.
Keycloak Integration: Analyzing the technical challenges and potential solutions for integrating a quantum-resistant algorithm into Keycloak's architecture.
Performance and Usability: Evaluating the impact of the new algorithm on Keycloak's performance and user experience.
Future-Proofing Keycloak: Discussing the long-term benefits of adopting a quantum-resistant approach and its role in securing identity and access management systems.
We welcome contributions from cryptography experts, security researchers, Keycloak developers, and anyone interested in exploring this critical topic. Let's work together to ensure the continued security of Keycloak in the quantum era.

Potential Discussion Points:

Which post-quantum cryptographic algorithms are most promising for password hashing?
What are the performance implications of different quantum-resistant algorithms?
How can we balance security with usability when implementing a new hashing algorithm?
What are the potential security risks of migrating to a new hashing algorithm?
How can we educate users about the importance of strong passwords in the quantum era?
Additional Considerations:

Standardization: Aligning with emerging standards for post-quantum cryptography can facilitate interoperability and adoption.
Backward Compatibility: Ensuring compatibility with existing password hashes to avoid disruption.
Key Length: Determining appropriate key lengths for quantum-resistant algorithms to provide adequate security.
Cost-Benefit Analysis: Weighing the investment in quantum-resistant technology against the potential risks of not adopting it.
By engaging in open and collaborative discussions, we can collectively address the challenges and opportunities presented by quantum computing and strengthen the security of Keycloak for the future.

**Comment by Ketec** [comment]
NIST has released first 3 standards for cryptography.
https://www.nist.gov/news-events/news/2024/08/nist-releases-first-3-finalized-post-quantum-encryption-standards
ML-KEM, ML-DSA, SLH-DSA

While these are not exactly international standards - perhaps the first approach here should be improving KC handling of cryptography to make introducing new algorithms easier. Less hard coded and more modular, standalone code.

These changes should also fix some of the kc limitations with current signatures/encryption keys being global and lacking per client or identity provider rotation.
